# Zomato Delivery Operations Analytics

## Project Objective
Zomato's operations team wants to understand what factors most impact 
delivery time and customer satisfaction across Indian cities. As a Data 
Analyst, the goal is to identify the biggest operational gaps and provide 
a prioritized, data-driven action plan for the operations team.

## Business Questions We Are Solving
1. Which city type has the worst delivery performance?
2. Does delivery person age or ratings affect delivery time?
3. Do weather and road traffic significantly increase delivery time — and by how much?
4. Which order types and vehicle types are associated with the most delays?
5. What is the single biggest lever the ops team should pull to reduce delivery time?

## Tools & Technologies
- Python 3.11
- Pandas — Data manipulation and analysis
- NumPy — Numerical computations
- Matplotlib — Data visualization
- Seaborn — Statistical data visualization
- SQLite3 — SQL-based analysis on structured data

## Dataset
- Source: Kaggle — Zomato Delivery Operations Analytics Dataset
- Size: 45,584 rows × 20 columns
- Each row represents one food delivery transaction

## Project Structure
- Phase 1 — Data Ingestion & Initial Inspection
- Phase 2 — Data Cleaning & Preprocessing
- Phase 3 — Exploratory Data Analysis (EDA)
- Phase 4 — Core Business Analysis
- Phase 5 — Data Visualization
- Phase 6 — SQL Analysis
- Phase 7 — Insights & Business Recommendations

## Phase 1 — Data Ingestion & Initial Inspection

**What is Data Ingestion?**
Data Ingestion is the process of loading raw data from its source into 
our working environment for analysis. This is always the first technical 
step in any data analytics project.

**What is Initial Inspection?**
Before doing anything with data, a professional analyst always performs 
an initial inspection — understanding the shape of the data, what each 
column represents, what data types are present, and whether there are 
any immediate red flags like missing values or incorrect formats.

**Why this matters:**
You cannot clean what you don't understand. You cannot analyse what you 
haven't inspected. This phase gives us the full picture of what we're 
working with before we touch a single value.

In [2]:
#---------------------------------------------------------------
# PHASE 1 — DATA INGESTION & INITIAL INSPECTION
#---------------------------------------------------------------
# Step 1: Import necessary libraries
import pandas as pd    # For data manipulation and analysis
import numpy as np     # For numerical operations 
import matplotlib.pyplot as plt   # Visualization 
import seaborn as sns  # Statistical Visualization 
import warnings
warnings.filterwarnings('ignore')

# Setting display options for better readability
pd.set_option('display.max_columns', None)
pd.set_option('display.width', 1000)

In [ ]:
# Step 1 -  Loading the dataset ──
df = pd.read_csv('/Users/mirajpatil/Documents/Zomato Delivery Analytics/Data/Zomato Dataset.csv')
print("✅ Dataset loaded successfully")

✅ Dataset loaded successfully


In [6]:
print(f"Shape: {df.shape[0]} rows × {df.shape[1]} columns")

Shape: 45584 rows × 20 columns


In [ ]:
# Step 2 -  Now lets do the data inspection by looking at the first few rows from Top and in next cell from bottom as well 
print('First 10 rows:')
print(df.head(10))

First 10 rows:
       ID Delivery_person_ID  Delivery_person_Age  Delivery_person_Ratings  Restaurant_latitude  Restaurant_longitude  Delivery_location_latitude  Delivery_location_longitude  Order_Date Time_Orderd Time_Order_picked Weather_conditions Road_traffic_density  Vehicle_condition Type_of_order   Type_of_vehicle  multiple_deliveries Festival           City  Time_taken (min)
0  0xcdcd      DEHRES17DEL01                 36.0                      4.2            30.327968             78.046106                   30.397968                    78.116106  12-02-2022       21:55             22:10                Fog                  Jam                  2         Snack        motorcycle                  3.0       No  Metropolitian                46
1  0xd987      KOCRES16DEL01                 21.0                      4.7            10.003064             76.307589                   10.043064                    76.347589  13-02-2022       14:55             15:05             Stormy        

In [12]:
print('Last 10 rows:')
print(df.tail(10))

Last 10 rows:
           ID Delivery_person_ID  Delivery_person_Age  Delivery_person_Ratings  Restaurant_latitude  Restaurant_longitude  Delivery_location_latitude  Delivery_location_longitude  Order_Date Time_Orderd Time_Order_picked Weather_conditions Road_traffic_density  Vehicle_condition Type_of_order   Type_of_vehicle  multiple_deliveries Festival           City  Time_taken (min)
45574  0x5193      MYSRES13DEL02                 36.0                      4.8            12.310972             76.659264                   12.440972                    76.789264  18-03-2022       21:10             21:20              Sunny                  Jam                  2        Drinks  electric_scooter                  1.0       No          Urban                29
45575  0xa333     CHENRES08DEL02                 37.0                      4.8            13.022394             80.242439                   13.042394                    80.262439  05-04-2022       09:35             09:50         Sandsto

In [13]:
# step 3 - Understanding data types and structure of data frame 
print('Data info:')
print(df.info())

Data info:
<class 'pandas.DataFrame'>
RangeIndex: 45584 entries, 0 to 45583
Data columns (total 20 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   ID                           45584 non-null  str    
 1   Delivery_person_ID           45584 non-null  str    
 2   Delivery_person_Age          43730 non-null  float64
 3   Delivery_person_Ratings      43676 non-null  float64
 4   Restaurant_latitude          45584 non-null  float64
 5   Restaurant_longitude         45584 non-null  float64
 6   Delivery_location_latitude   45584 non-null  float64
 7   Delivery_location_longitude  45584 non-null  float64
 8   Order_Date                   45584 non-null  str    
 9   Time_Orderd                  43853 non-null  str    
 10  Time_Order_picked            45584 non-null  str    
 11  Weather_conditions           44968 non-null  str    
 12  Road_traffic_density         44983 non-null  str    
 13  Vehicle_conditio

In [14]:
# step 4 - Statistical summary of numerical values 
print('statistical summary:')
print(df.describe())

statistical summary:
       Delivery_person_Age  Delivery_person_Ratings  Restaurant_latitude  Restaurant_longitude  Delivery_location_latitude  Delivery_location_longitude  Vehicle_condition  multiple_deliveries  Time_taken (min)
count         43730.000000             43676.000000         45584.000000          45584.000000                45584.000000                 45584.000000       45584.000000         44591.000000      45584.000000
mean             29.566911                 4.633774            17.017948             70.229684                   17.465480                    70.844161           1.023385             0.744635         26.293963
std               5.815064                 0.334744             8.185674             22.885575                    7.335562                    21.120578           0.839055             0.572510          9.384298
min              15.000000                 1.000000           -30.905562            -88.366217                    0.010000                 

In [21]:
# step 5 - checking Unique values in categorical column

categorical_cols = ['Weather_conditions','Road_traffic_density','Type_of_order','Type_of_vehicle','Festival','City']

for col in categorical_cols:
    print(f'\n{col}: {df[col].nunique()} unique values')
    print(df[col].value_counts())




Weather_conditions: 6 unique values
Weather_conditions
Fog           7653
Stormy        7584
Cloudy        7533
Sandstorms    7494
Windy         7422
Sunny         7282
Name: count, dtype: int64

Road_traffic_density: 4 unique values
Road_traffic_density
Low       15476
Jam       14139
Medium    10945
High       4423
Name: count, dtype: int64

Type_of_order: 4 unique values
Type_of_order
Snack     11530
Meal      11456
Drinks    11321
Buffet    11277
Name: count, dtype: int64

Type_of_vehicle: 4 unique values
Type_of_vehicle
motorcycle          26429
scooter             15273
electric_scooter     3814
bicycle                68
Name: count, dtype: int64

Festival: 2 unique values
Festival
No     44460
Yes      896
Name: count, dtype: int64

City: 3 unique values
City
Metropolitian    34087
Urban            10133
Semi-Urban         164
Name: count, dtype: int64


## Phase 1 — Summary of Initial Inspection Findings

| Finding | Column | Action Required |
|---|---|---|
| Ages below 18 found | Delivery_person_Age | Remove rows where age < 18 |
| Rating value of 6.0 found | Delivery_person_Ratings | Cap ratings at 5.0 |
| Negative coordinates found | Lat/Long columns | Drop coordinate columns entirely |
| Misspelled city name | City | Rename Metropolitian → Metropolitan |
| Semi-Urban has only 164 rows | City | Drop Semi-Urban rows |
| Bicycle has only 68 rows | Type_of_vehicle | Drop Bicycle rows |
| Time_Orderd has 1731 nulls | Time_Orderd | Drop column — not needed |
| Multiple columns have nulls | Age, Ratings, Weather, Traffic, Festival, City | Handle in cleaning phase |

**Target Variable Status:** Time_taken (min) — 45,584 non-null — Clean ✅